### ISSUE TITLE
sysusers.d man page states wrong lexicographically order

### ISSUE THREAD

**Issue by grazzolini** [systemd/systemd#6639]
### Submission type

 - Bug report

### systemd version the issue has been seen with

234

### Used distribution

Arch Linux

### In case of bug report: Expected behaviour you didn't see

The sysusers.d man page states: "If multiple files specify the same option, the entry in the file with the lexicographically latest name will take precedence."

### In case of bug report: Unexpected behaviour you saw

The files are processed in the reverse order of what the man page states, the entry in the first file being the one used and the last one being discarded.

### In case of bug report: Steps to reproduce the problem

1) Create two files with conflicting entries in any of the directories systemd expect sysusers files, /etc, /run or /usr/lib.
2) Run systemd-sysusers
3) The first entry in the first file will be used and the second entry will be discarded.

I have come across this issue when investigating a bug report: https://bugs.archlinux.org/task/54943

Arch Linux has been adding the kvm group when qemu is installed but with a fixed group id (GID) of 78. After systemd 234 was deployed, users which did not had qemu installed, got a kvm group created with a random GID. Since all of qemu/libvirt configuration on Arch Linux expects the fixed GID, their libvirt/qemu installation was non functional. Users who installed qemu before 234, are not affected by this.

After understanding the issue, I have created a basic-arch.conf file which has the fixed GID and overrides systemd configuration. But, lexically, this file is before systemd's provided basic.conf, not after. Either the man page needs to get changed or systemd-sysuser itself. From what I've seen on other systemd tools, usually the first file wins, I only saw one other case where it is stated that the last one wins.

**Comment by yuwata** [comment]
I've tested with the following three files in `/etc/sysusers.d`
```
# 00-foo.conf
g foo     900     -            -
# 01-foo-aaa.conf
g foo     902     -            -
# 01-foo.conf
g foo     -     -            -
```
If all three files are exist, then the group foo is created with gid 900. Without 00-foo.conf, that is, only 01-foo-aaa.conf and 01-foo.conf exist, then the group is created with gid 902. So, it seems no problem. Or, I may confuse your problem.

**Comment by yuwata** [comment]
Sysusers ignores an entry if the corresponding user or group already exists. Do you create basic-arch.conf before the group kvm is created?

**Comment by grazzolini** [comment]
@yuwata 
The issue is that the behavior does not match the man page, that's all. The man page states that the **last** filename will be used, not the first. I actually prefer how it works right now, matching the first filename. Just that the man page needs fixing. I have not provided a patch, because I don't know if the man page will be fixed, or if systemd-sysusers will get patched to switch the behavior. If it's just a man page issue, I'd be glad to provide a patch.

**Comment by yuwata** [comment]
Thanks. I think the man page should be fixed. This behavior is the same as the tmpfiles and similar to [Match] section in the .network file.

**Comment by grazzolini** [comment]
@yuwata 
I went looking into writing a PR for this but it seems that this is derived from standard-conf.xml. And that file is include by quite a few other files. I don't know if all those tools follows the same precedence, but if the normal precedence on systemd is first to last, then I guess that changing that file will fix the man pages of all of those that includes it.

**Comment by yuwata** [comment]
So, sysusers.d.xml should not to include standard-conf.xml. I think the second and third paragraph of the section "Configuration Format" in tmpfiles.d.xml can be a good reference.
Please do not edit standard-conf.xml.

**Comment by grazzolini** [comment]
@yuwata 
Sure, I would not change the documentation without knowing for sure that all the man pages that includes it, actually work in that way. I'll try to prepare a patch for this today and send a PR.

**Comment by stephankn** [comment]
I observed a similar issue with systemd-timesyncd. As I faced it with Ubuntu 16.04 LTS having systemd 229 I can't file a bug against this older software. 

It sounds quite likely that the documentation bug also applies there.

If I place multiple configs in the timesyncd.conf.d folder the one with the **first** filename is used, as opposed to the last one mentioned in the manpage.

Would be great if you could do a quick check and adjust the documentation there as well. The code to parse config seems to be shared among all utilities, so they are all affected by the same issue.

**Comment by yuwata** [comment]
@stephankn I do not reproduce your problem. I've tested with multiple drop-in configs but all settings seems correctly read.
I note that the man page systemd-timesyncd(8) states that
> The systemd-timesyncd service specifically implements only SNTP.

So, timesyncd uses only one, usually the first, NTP server in listed servers.
If you want to use the NTP server listed in the second or later position, I propose to use `NTP=` to clear currently listed servers. E.g.,
```
# /etc/systemd/timesyncd.conf.d/a.conf
[Time]
NTP=ntp1.xxxx.org
# /etc/systemd/timesyncd.conf.d/b.conf
NTP=
NTP=ntp2.xxxx.org
```
Then you can use `ntp2.xxxx.org` as an NTP server.

**Comment by stephankn** [comment]
@yuwata I am thankful you try to help me. I wonder how to best proceed as the issue might be different to what is originally tracked here. My impression was that the code to parse config files existed only once in the systemd codebase. So it would not matter whether it reads timesyncd config or sysusers config. 
And with it being confirmed that documentation was wrong and the lexically first file wins this was in line with my observations regarding timesyncd. So I added the comment here.

I wonder how to best reproduce it with a recent version of systemd to be able to create a dedicated bug-report. With Ubuntu 16.04 LTS and 229 it can be seen as:

```
# grep NTP= /etc/systemd/timesyncd.conf.d/*
/etc/systemd/timesyncd.conf.d/hetzner.conf:NTP=ntp1.hetzner.de ntp2.hetzner.com ntp3.hetzner.net
/etc/systemd/timesyncd.conf.d/zz-default.conf:NTP=0.de.pool.ntp.org time.google.com ntp.ubuntu.com

# systemctl restart systemd-timesyncd

# journal lists the server from the *first* file used.
Oct 06 00:22:05 kontiki systemd-timesyncd[30096]: Synchronized to time server [2a01:4f8:0:a0a1::2:1]:123 (ntp1.hetzner.de).

# mv zz-default.conf 00default.conf
# grep NTP= /etc/systemd/timesyncd.conf.d/*
/etc/systemd/timesyncd.conf.d/00default.conf:NTP=0.de.pool.ntp.org time.google.com ntp.ubuntu.com
/etc/systemd/timesyncd.conf.d/hetzner.conf:NTP=ntp1.hetzner.de ntp2.hetzner.com ntp3.hetzner.net

# systemctl restart systemd-timesyncd
# journal lists the server from the *first* file used.
Oct 06 00:28:00 kontiki systemd-timesyncd[30617]: Synchronized to time server 82.100.248.10:123 (0.de.pool.ntp.org).
```

So with version 229 clearly the first file wins and the entry in the manual is wrong.
man timesyncd.conf states:

    If multiple files specify the same option, the entry in the file with the lexicographically latest name takes precedence.

This is clearly not true as at least on 229 the *first* file wins, which is also in line with the documentation change done above.

So how to proceed? Are you reproducing it in the same way? Shall I open a dedicated issue, despite the version being older?

**Comment by keszybz** [comment]
OK, it seems that the documentation is *still* unclear. I'll push another PR to clarify this.

**Comment by stephankn** [comment]
@keszybz Ah, it appends options, instead of overwriting them. Many thanks for clarifying this important detail for list options.

Is there an easy way to figure out what is the actual content of the NTP setting, is it? systemctl show does not list it. Would have made it probably much clearer what actually happend when seeing the full list and not just the first entry being used.

So if it is easily possible to list the actual content of list options, mentioning in the man-page how to read it certainly helps others to debug configuration issues.

**Comment by keszybz** [comment]
It will switch to the next entry in the list if the first server does not respond. But no, I don't think there's a way to show this information. For some daemons (systemd, systemd-resolved) we have ways to dump internal state, but not for all of them.

**Comment by yuwata** [comment]
@stephankn 
> Is there an easy way to figure out what is the actual content of the NTP setting, is it?

It may not be a good idea, but you can confirm the setting by creating the following drop-in config.
```
# /etc/systemd/system/systemd-timesyncd.service.d/debug.conf
[Service]
Environment=SYSTEMD_LOG_LEVEL=debug
```
